packages/astro/src/content/runtime.ts

```diff
@@ -319,9 +334,89 @@ type RenderResult = {
 	remarkPluginFrontmatter: Record<string, any>;
 };
 
+const IMAGE_REGEX = /__ASTRO_IMAGE_="([^"]+)"/g;
+
+async function updateImageReferencesInBody(html: string, fileName: string) {
+	// @ts-expect-error Virtual module
+	const { default: imageAssetMap } = await import('astro:asset-imports');
+
+	const imageObjects = new Map<string, GetImageResult>();
+
+	// @ts-expect-error Virtual module resolved at runtime
+	const { getImage } = await import('astro:assets');
+
+	for (const match of html.matchAll(IMAGE_REGEX)) {
+		const imagePath = match[1];
+		const decodedImagePath = JSON.parse(imagePath.replace(/&#x22;/g, '"'));
+		const id = imageSrcToImportId(decodedImagePath.src, fileName);
+
+		if (!id || imageObjects.has(id)) {
+			continue;
+		}
+
+		const imported = imageAssetMap.get(id);
+		if (!imported) {
+			continue;
+		}
+		const image: GetImageResult = await getImage({ src: imported });
+		imageObjects.set(id, image);
+	}
+
+	return html.replaceAll(IMAGE_REGEX, (full, imagePath) => {
+		const decodedImagePath = JSON.parse(imagePath.replace(/&#x22;/g, '"'));
+		const id = imageSrcToImportId(decodedImagePath.src, fileName);
+
+		if (!id) {
+			return full;
+		}
+
+		const image = imageObjects.get(id);
+
+		if (!image) {
+			return full;
+		}
+
+		return Object.entries({
+			...image.attributes,
+			src: image.src,
+			srcset: image.srcSet.attribute,
+			alt: decodedImagePath.alt,
+		})
+			.map(([key, value]) => (value ? `${key}=${JSON.stringify(String(value))}` : ''))
+			.join(' ');
+	});
+}
+
+function updateImageReferencesInData<T extends Record<string, unknown>>(
+	data: T,
+	fileName: string,
+	imageAssetMap: Map<string, ImageMetadata>
+): T {
+	return new Traverse(data).map(function (ctx, val) {
```

Comment:
I benchmarked this with 10k entries and it didn't add any noticable time to the processing.